cli/start_file_writer.py

```diff
@@ -0,0 +1,221 @@
+    )
+    fw_parser.add_argument(
+        "-t",
+        "--topic",
+        metavar="consume_topic",
+        type=str,
+        required=True,
+        help="Name of the Kafka topic to be consumed.",
+    )
+    fw_parser.add_argument(
+        "--timeout",
+        metavar="ack_timeout",
+        type=float,
+        default=5,
+        help="How long to wait for timeout on acknowledgement.",
+    )
+    fw_parser.add_argument(
+        "--stop",
+        metavar="stop_writing",
+        type=float,
+        help="How long the file will be written.",
+    )
+
+    args = fw_parser.parse_args()
+
+    return args
+
+
+def file_writer(args: argparse.Namespace) -> None:
+    write_job = prepare_write_job(args)
+    start_time, timeout = start_write_job(write_job)
+
+    if args.stop:
+        stop_write_job(args.stop, start_time, timeout)
+
+    inform_status()
+
+
+def start_write_job(write_job: WriteJob) -> Tuple[datetime, float]:
+
+    start_handler = JOB_HANDLER.start_job(write_job)
+    timeout = int(current_time()) + ACK_TIMEOUT
+    start_time = datetime.now()
+    while not start_handler.is_done():
+        if int(current_time()) > timeout:
+            raise ValueError("Timeout.")
+    return start_time, timeout
+
+
+def stop_write_job(stop: float, start_time: datetime, timeout: float) -> None:
+
+    stop_time = start_time + timedelta(seconds=stop)
+    stop_handler = JOB_HANDLER.set_stop_time(stop_time)
+    while not stop_handler.is_done() and not JOB_HANDLER.is_done():
+        if int(current_time()) > timeout:
+            raise ValueError("Timeout.")
+
+
+def stop_write_job_now() -> None:
+    while JOB_HANDLER.get_state() == JobState.WRITING:
+        JOB_HANDLER.stop_now()
+        time.sleep(1)
+        if JOB_HANDLER.get_state() == JobState.DONE:
+            print("FileWriter successfully stopped.")
+    sys.exit()
+
+
+def prepare_write_job(args: argparse.Namespace) -> WriteJob:
+    global JOB_HANDLER
+    global ACK_TIMEOUT
+
+    file_name = args.filename
+    host = args.broker
+    topic = args.topic
+    config = args.config
+    ACK_TIMEOUT = args.timeout
+    command_channel = WorkerCommandChannel(f"{host}/{topic}")
+    JOB_HANDLER = JobHandler(worker_finder=command_channel)
+    with open(config, "r") as f:
+        nexus_structure = f.read()
+    write_job = WriteJob(
+        nexus_structure,
+        file_name,
+        host,
+        datetime.now(),
+    )
+    return write_job
+
+
+def inform_status() -> None:
+    if JOB_HANDLER.get_state() == JobState.WRITING:
+        print("Writing.", end="", flush=True)
+    while JOB_HANDLER.get_state() == JobState.WRITING:
+        print(".", end="", flush=True)
+        time.sleep(1)
+        if JOB_HANDLER.get_state() == JobState.DONE:
+            print("[DONE]")
+
+
+def validate_namespace(args: argparse.Namespace) -> None:
+    argument_list = [args.filename, args.config, args.broker, args.topic]
+    for arg in argument_list:
+        is_empty(arg)
```

Comment:
This check is probably redundant. argsparse will check it anyway.

Reply:
Nope. It accepts empty.

Reply:
Because the type is string, and empty or space is type string. For argparse there is no problem. It will just rise because the setting will probably fail to start FW.

Reply:
```
import argparse
def cli_parser() -> argparse.Namespace:
    """
    Parser for the command line interface.
    """

    fw_parser = argparse.ArgumentParser(
        fromfile_prefix_chars="@", description="FileWriter Starter"
    )

    fw_parser.add_argument(
        "-f",
        "--filename",
        metavar="filename",
        type=str,
        required=True,
        help="Name of the output file, e.g., `<filename>.nxs`.",
    )
    fw_parser.add_argument(
        "-c",
        "--config",
        metavar="json_config",
        type=str,
        required=True,
        help="Path to JSON config file.",
    )
    args = fw_parser.parse_args()

    return args

if __name__ == "__main__":
    cli_args = cli_parser()
```

I tried running this part of code with
`python test.py -f file -c   `
it throws error

```
usage: test.py [-h] -f filename -c json_config
test.py: error: argument -c/--config: expected one argument

```
So in a way it is not allowing space.

Reply:
you should provide the config after `-c`

Reply:
as it says in usage.

Reply:
`python cli.py -f filename -c config`

Reply:
no it says the argument is missing. Try with `" "` and it will pass.

Reply:
Yes, that was the point. So argsparse does not allow empty or spaces for required arguments. Therefore checking `is_empty` in validation seems redundant. No?

Reply:
So it interprets `x` and `"x"` same. But no argument is just no argument, nothing to do with space.

Reply:
But as I said, as a user I can also put `"x.json"` instead of `x.json`. You see? Then `" "` will be a problem.

Reply:
So these two are equivalent:

- `python cli.py -f filename.nxs -c ali.json`
- `python cli.py -f "filename.nxs" -c "ali.json"`

Reply:
and this will not throw any error:

- `python cli.py -f " " -c ""`

Reply:
understood. for files anyway you check for extensions and all. For brokers and topic some drunk user can provide " " .

Reply:
I just realised the frikin thing accepts a name like `.nxs` :S So I need to improve that check.